Modules/_sqlite/blob.c

```diff
@@ -0,0 +1,643 @@
+    char *raw_buffer;
+    int rc;
+
+    buffer = PyBytes_FromStringAndSize(NULL, read_length);
+    if (!buffer) {
+        return NULL;
+    }
+    raw_buffer = PyBytes_AS_STRING(buffer);
+
+    Py_BEGIN_ALLOW_THREADS
+    rc = sqlite3_blob_read(self->blob, raw_buffer, read_length, self->offset);
+    Py_END_ALLOW_THREADS
+
+    if (rc != SQLITE_OK){
+        Py_DECREF(buffer);
+        /* For some reason after modifying blob the
+           error is not set on the connection db. */
+        if (rc == SQLITE_ABORT) {
+            PyErr_SetString(pysqlite_OperationalError,
+                            "Cannot operate on modified blob");
+        } else {
+            _pysqlite_seterror(self->connection->db, NULL);
+        }
+        return NULL;
+    }
+    return buffer;
+}
+
+
+PyObject* pysqlite_blob_read(pysqlite_Blob *self, PyObject *args)
+{
+    int read_length = -1;
+    PyObject *buffer;
+
+    if (!PyArg_ParseTuple(args, "|i", &read_length)) {
+        return NULL;
+    }
+
+    if (!pysqlite_check_blob(self)) {
+        return NULL;
+    }
+
+    if (read_length < 0) {
+        /* same as file read. */
+        read_length = self->length;
+    }
+
+    /* making sure we don't read more then blob size */
+    if (read_length > self->length - self->offset) {
+        read_length = self->length - self->offset;
+    }
+
+    buffer = inner_read(self, read_length, self->offset);
+
+    if (buffer != NULL) {
+        /* update offset on sucess. */
+        self->offset += read_length;
+    }
+
+    return buffer;
+};
+
+static int write_inner(pysqlite_Blob *self, const void *buf, Py_ssize_t len, int offset)
+{
+    int rc;
+
+    Py_BEGIN_ALLOW_THREADS
+    rc = sqlite3_blob_write(self->blob, buf, len, offset);
+    Py_END_ALLOW_THREADS
+    if (rc != SQLITE_OK) {
+        /* For some reason after modifying blob the
+        error is not set on the connection db. */
+        if (rc == SQLITE_ABORT) {
+            PyErr_SetString(pysqlite_OperationalError,
+                            "Cannot operate on modified blob");
+        } else {
+            _pysqlite_seterror(self->connection->db, NULL);
+        }
+        return -1;
+    }
+    return 0;
+}
+
+
+PyObject* pysqlite_blob_write(pysqlite_Blob *self, PyObject *data)
+{
+    Py_buffer data_buffer;
+    int rc;
+
+    if (PyObject_GetBuffer(data, &data_buffer, PyBUF_SIMPLE) < 0) {
+        return NULL;
+    }
+
+    if (data_buffer.len > INT_MAX) {
+        PyErr_SetString(PyExc_OverflowError,
+                        "data longer than INT_MAX bytes");
+        PyBuffer_Release(&data_buffer);
+        return NULL;
+    }
+
+    if (!pysqlite_check_blob(self)) {
+        PyBuffer_Release(&data_buffer);
+        return NULL;
+    }
+
+    /* TODO: throw better error on data bigger then blob. */
+
+    rc = write_inner(self, data_buffer.buf, data_buffer.len, self->offset);
+
+    if (rc == 0) {
+        self->offset += (int)data_buffer.len;
+        Py_RETURN_NONE;
```

Comment:
It seems that `data_buffer` is leaked here.

Reply:
Hi @ZackerySpytz. I agree that there is a leak here.
This patch exist since January 2016 and I kind of given up on it ever going into CPython as there is no core developer that works on sqlite. I would recommend you to use apsw that support this feature. In case any core developer would be in interested in working on this I would gladly fix any needed changes.